### DISCUSSION TITLE
dynamicGallery.refresh() does not allow you to add new video sources

### DISCUSSION THREAD

**Discussion by swstyles** [sachinchoolur/lightGallery#1074] [General]
On the first init of dynamicGallery, all instances of videos and images load and play fine. But when you add more videos to the object, you receive the "Oops.... Failed to Load" message.

In the code below, I am clearing all the current gallery values and only adding new values including a mix of video and images. Only the images load. The video thumbnail displays but the video fails to load.
const updatedDynamicElements = [...newImages];
dynamicGallery.refresh(updatedDynamicElements);

TIA

**Comment by sachinchoolur** [comment]
Thank you for reporting the issue. 
Fixed in version [2.2.7](https://github.com/sachinchoolur/lightGallery/releases/tag/2.1.7)

**Reply by swstyles** [reply]
Works great now. Thanks for the quick fix.

**Comment by swstyles** [comment]
Instead of starting a new thread I will continue to post here since this new issue is related.

When I hard code the dynamicEl object values, every thing works fine after refresh(). As I am moving towards making my lightgallery dynamic I am running into an issue when using fetch() and JSON files. After refresh for some reason the first NEW loaded image is not showing. The thumbnail refreshes but the large image doesn't. Instead it shows the first large image of the previously loaded gallery. If you collapse the gallery and click on the same initial thumbnail again, the correct large image does appear.

You can see the issue here:
http://talent.leightonagency.com/lg_la/sample1.html 

Thanks

**Reply by sachinchoolur** [reply]
Hey @swstyles,

I think this is happening because you are opening gallery before calling the refresh method.

```js
function getImages(link){
  fetch(link)
    .then(response => response.json())
    .then(data => {
      dynamicGallery.openGallery();
      const updatedDynamicElements = [...data];
      dynamicGallery.refresh(updatedDynamicElements);

      // To open gallery after updating slides,
      dynamicGallery.openGallery();
      console.log(dynamicGallery.galleryItems[0].src);
    });
}
```

Could you please try removing the first `dynamicGallery.openGallery();` method

```js
function getImages(link){
  fetch(link)
    .then(response => response.json())
    .then(data => {
      const updatedDynamicElements = [...data];
      dynamicGallery.refresh(updatedDynamicElements);

      // To open gallery after updating slides,
     // You can also try calling the below method in setTimeout if required
      dynamicGallery.openGallery();
    });
}
```

In case, if the issue persists, please provide an isolated demo. I'll fix the issue and send you back

**Reply by swstyles** [reply]
I was playing around with that trying to see if adding openGallery() twice would fix the issue but it didn't. I went ahead and removed the first instance. You can see that the problem still exists.

**Reply by sachinchoolur** [reply]
Just tried it on my local server with a similar setup. Everything seems to be working fine for me.

Is it possible for you to provide an isolated demo? If you don't want to share it here you can send me an email to contact@lightgalleryjs.com. I'll try to resolve the issue ASAP

**Reply by swstyles** [reply]
If you open the developer console and click on the first top left thumbnail, you will see the YouTube url that should appear in the main image area. The thumbnail appears as it should but instead of a video you get an image that was loaded from init. The image shows the letters la.

http://talent.leightonagency.com/lg_la/sample1.html

Could I see how you implemented fetch() and refresh() in your test code?

**Reply by sachinchoolur** [reply]
I'm sorry, I had reverted all my changes when I found it was working. 
Yes, I looked into the source code and was able to reproduce the issue. If you could enable "Cross-Origin" for some time for the [json files](http://talent.leightonagency.com/lg_la/2.json), I can test it with the same source code itself

**Reply by swstyles** [reply]
I am not sure how to enable CORS on my server.

**Reply by swstyles** [reply]
looks like I can add the headers in the web.config file.
Can you provide your local server address so I can add it to the allow list?

Thanks

**Reply by swstyles** [reply]
I went ahead and opened CORS for you. I also fully pathed all links instead of using relative so you don't receive any 404 errors in your local view.

I appreciate all your help on this.

**Reply by sachinchoolur** [reply]
Thanks for making the changes.

The issue was happening due to the following lines 

```js
$dynamicGallery.addEventListener("click", () => {
  dynamicGallery.openGallery(0);
});
```

The gallery is getting opened when the user clicks on any image as it was added to the container. So the  `openGallery()` method inside `getImages` function is doesn't have any effect as the gallery is already in the open state.

let me know if you face any issues

**Reply by swstyles** [reply]
Wow, I just removed that block of code and it works fine now. It's weird because when I hard coded the dynamicEl values it worked fine. It only stopped working when I used imported JSON values.

Thank you for helping me get this working.

Have a good one!
